Build failure with fmt 9.0.0

### Describe the bug

vowpal_wabbit 9.2.0 fails to build with fmt 9.0.0.

### How to reproduce

Build vowpal_wabbit with fmt 9.0.0.

Or, if you already have a Homebrew installation handy:
```
cd "$(brew --repository homebrew/core)"
gh pr checkout 105132
brew install --build-from-source fmt
brew install --build-from-source --verbose vowpal_wabbit
```

### Version

9.2.0

### OS

macOS

### Language

C++

### Additional context

We're trying to update Homebrew's fmt at https://github.com/Homebrew/homebrew-core/pull/105132. While rebuilding dependents, we encountered the following build failure for vowpal_wabbit:

```
  /opt/homebrew/include/fmt/core.h:1732:3: error: static_assert failed due to requirement 'formattable' "Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt"
    static_assert(
    ^
  /opt/homebrew/include/fmt/core.h:1753:10: note: in instantiation of function template specialization 'fmt::detail::make_value<fmt::basic_format_context<fmt::appender, char>, nonstd::sv_lite::basic_string_view<char> &>' requested here
    return make_value<Context>(val);
           ^
  /opt/homebrew/include/fmt/core.h:1875:23: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::basic_format_context<fmt::appender, char>, fmt::detail::type::custom_type, nonstd::sv_lite::basic_string_view<char> &, 0>' requested here
          data_{detail::make_arg<
                        ^
  /opt/homebrew/include/fmt/core.h:1894:10: note: in instantiation of function template specialization 'fmt::format_arg_store<fmt::basic_format_context<fmt::appender, char>, nonstd::sv_lite::basic_string_view<char>>::format_arg_store<nonstd::sv_lite::basic_string_view<char> &>' requested here
    return {std::forward<Args>(args)...};
           ^
  /opt/homebrew/include/spdlog/logger.h:90:9: note: in instantiation of function template specialization 'spdlog::logger::log_<nonstd::sv_lite::basic_string_view<char> &>' requested here
          log_(loc, lvl, fmt, std::forward<Args>(args)...);
          ^
  /opt/homebrew/include/spdlog/logger.h:96:9: note: in instantiation of function template specialization 'spdlog::logger::log<nonstd::sv_lite::basic_string_view<char> &>' requested here
          log(source_loc{}, lvl, fmt, std::forward<Args>(args)...);
          ^
  /opt/homebrew/include/spdlog/logger.h:164:9: note: in instantiation of function template specialization 'spdlog::logger::log<nonstd::sv_lite::basic_string_view<char> &>' requested here
          log(level::warn, fmt, std::forward<Args>(args)...);
          ^
  /tmp/vowpal-wabbit-20220711-46392-19otvmm/vowpal_wabbit-9.1.0/vowpalwabbit/io/logger.h:175:74: note: in instantiation of function template specialization 'spdlog::logger::warn<nonstd::sv_lite::basic_string_view<char> &>' requested here
        if (_location == output_location::compat) { _spdlog_stdout_logger->warn(fmt, std::forward<Args>(args)...); }
                                                                           ^
  /tmp/vowpal-wabbit-20220711-46392-19otvmm/vowpal_wabbit-9.1.0/vowpalwabbit/io/logger.h:317:19: note: in instantiation of function template specialization 'VW::io::details::logger_impl::out_warn<char [43], nonstd::sv_lite::basic_string_view<char> &>' requested here
      _logger_impl->out_warn(fmt, std::forward<Args>(args)...);
                    ^
  /tmp/vowpal-wabbit-20220711-46392-19otvmm/vowpal_wabbit-9.1.0/vowpalwabbit/parse_primitives.h:117:12: note: in instantiation of function template specialization 'VW::io::logger::out_warn<char [43], nonstd::sv_lite::basic_string_view<char> &>' requested here
      logger.out_warn("'{}' is not a good float, replacing with 0", s);
             ^
```

Build logs are available at https://github.com/Homebrew/homebrew-core/runs/7277894282?check_suite_focus=true. The error starts at https://github.com/Homebrew/homebrew-core/runs/7277894282?check_suite_focus=true#step:6:10145.

I don't do C++, so apologies if the issue is not actually in vowpal_wabbit. I'm happy to report this upstream if the problem lies there.

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -6,13 +6,5 @@
             ],
             "url": "./test/vwtest.schema.json"
         }
-    ],
-    "files.associations": {
-        "*.tcc": "cpp",
-        "string_view": "cpp",
-        "regex": "cpp",
-        "alignedvector3": "cpp",
-        "array": "cpp",
-        "string": "cpp"
-    }
+    ]
 }

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -6,5 +6,13 @@
             ],
             "url": "./test/vwtest.schema.json"
         }
-    ]
+    ],
+    "files.associations": {
+        "*.tcc": "cpp",
+        "string_view": "cpp",
+        "regex": "cpp",
+        "alignedvector3": "cpp",
+        "array": "cpp",
+        "string": "cpp"
+    }
 }

--- a/python/vowpalwabbit/pyvw.py
+++ b/python/vowpalwabbit/pyvw.py
@@ -482,7 +482,7 @@ class Workspace(pylibvw.vw):
         self.finish()
 
     def parse(
-        self, str_ex: Union[str, List["Example"]], labelType: Optional[Union[int, LabelType]] = None
+        self, str_ex, labelType: Optional[Union[int, LabelType]] = None
     ) -> Union["Example", List["Example"]]:
         """Returns a collection of examples for a multiline example learner or
         a single example for a single example learner.

--- a/python/vowpalwabbit/pyvw.py
+++ b/python/vowpalwabbit/pyvw.py
@@ -482,7 +482,7 @@ class Workspace(pylibvw.vw):
         self.finish()
 
     def parse(
-        self, str_ex, labelType: Optional[Union[int, LabelType]] = None
+        self, str_ex: Union[str, List["Example"]], labelType: Optional[Union[int, LabelType]] = None
     ) -> Union["Example", List["Example"]]:
         """Returns a collection of examples for a multiline example learner or
         a single example for a single example learner.

--- a/vowpalwabbit/core/include/vw/core/action_score.h
+++ b/vowpalwabbit/core/include/vw/core/action_score.h
@@ -78,4 +78,12 @@ constexpr inline bool action_score_compare_gt(
 
 std::string to_string(
     const ACTION_SCORE::action_scores& action_scores_or_probs, int decimal_precision = DEFAULT_FLOAT_PRECISION);
+
+namespace model_utils
+{
+size_t read_model_field(io_buf& io, ACTION_SCORE::action_score& a_s);
+
+size_t write_model_field(
+    io_buf& io, const ACTION_SCORE::action_score a_s, const std::string& upstream_name, bool text);
+}  // namespace model_utils
 }  // namespace VW

--- a/vowpalwabbit/core/include/vw/core/action_score.h
+++ b/vowpalwabbit/core/include/vw/core/action_score.h
@@ -83,7 +83,6 @@ namespace model_utils
 {
 size_t read_model_field(io_buf& io, ACTION_SCORE::action_score& a_s);
 
-size_t write_model_field(
-    io_buf& io, const ACTION_SCORE::action_score a_s, const std::string& upstream_name, bool text);
+size_t write_model_field(io_buf& io, const ACTION_SCORE::action_score a_s, const std::string& upstream_name, bool text);
 }  // namespace model_utils
 }  // namespace VW

--- a/vowpalwabbit/core/include/vw/core/learner.h
+++ b/vowpalwabbit/core/include/vw/core/learner.h
@@ -26,6 +26,7 @@
 #undef VW_DEBUG_LOG
 #define VW_DEBUG_LOG vw_dbg::learner
 
+#include "vw/core/vw_string_view_fmt.h"
 #include "vw/common/future_compat.h"
 #include "vw/core/example.h"
 #include "vw/core/label_type.h"

--- a/vowpalwabbit/core/include/vw/core/learner.h
+++ b/vowpalwabbit/core/include/vw/core/learner.h
@@ -27,6 +27,7 @@
 #define VW_DEBUG_LOG vw_dbg::learner
 
 #include "vw/core/vw_string_view_fmt.h"
+
 #include "vw/common/future_compat.h"
 #include "vw/core/example.h"
 #include "vw/core/label_type.h"

--- a/vowpalwabbit/core/include/vw/core/parse_primitives.h
+++ b/vowpalwabbit/core/include/vw/core/parse_primitives.h
@@ -3,13 +3,14 @@
 // license as described in the file LICENSE.
 #pragma once
 
+#include "vw/core/vw_string_view_fmt.h"
+
 #include "fast_pow10.h"
 #include "hashstring.h"
 #include "vw/common/future_compat.h"
 #include "vw/common/string_view.h"
 #include "vw/core/v_array.h"
 #include "vw/io/logger.h"
-#include "vw/core/vw_string_view_fmt.h"
 
 #include <cmath>
 #include <cstdint>

--- a/vowpalwabbit/core/include/vw/core/parse_primitives.h
+++ b/vowpalwabbit/core/include/vw/core/parse_primitives.h
@@ -9,6 +9,7 @@
 #include "vw/common/string_view.h"
 #include "vw/core/v_array.h"
 #include "vw/io/logger.h"
+#include "vw/core/vw_string_view_fmt.h"
 
 #include <cmath>
 #include <cstdint>

--- a/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
+++ b/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
@@ -19,8 +19,13 @@ namespace fmt
 template <>
 struct formatter<VW::string_view> : formatter<fmt::string_view>
 {
+  #if FMT_VERSION >= 90000
   template <typename FormatContext>
   auto format(const VW::string_view& sv, FormatContext& ctx) const -> decltype(ctx.out())
+#else
+  template <typename FormatContext>
+  auto format(const VW::string_view& sv, FormatContext& ctx) -> decltype(ctx.out())
+#endif
   {
     return formatter<fmt::string_view>::format({sv.data(), sv.size()}, ctx);
   }

--- a/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
+++ b/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
@@ -20,7 +20,7 @@ template <>
 struct formatter<VW::string_view> : formatter<fmt::string_view>
 {
   template <typename FormatContext>
-  auto format(const VW::string_view& sv, FormatContext& ctx) -> decltype(ctx.out())
+  auto format(const VW::string_view& sv, FormatContext& ctx) const -> decltype(ctx.out())
   {
     return formatter<fmt::string_view>::format({sv.data(), sv.size()}, ctx);
   }

--- a/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
+++ b/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
@@ -19,13 +19,8 @@ namespace fmt
 template <>
 struct formatter<VW::string_view> : formatter<fmt::string_view>
 {
-#if FMT_VERSION >= 90000
   template <typename FormatContext>
   auto format(const VW::string_view& sv, FormatContext& ctx) const -> decltype(ctx.out())
-#else
-  template <typename FormatContext>
-  auto format(const VW::string_view& sv, FormatContext& ctx) -> decltype(ctx.out())
-#endif
   {
     return formatter<fmt::string_view>::format({sv.data(), sv.size()}, ctx);
   }

--- a/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
+++ b/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
@@ -19,7 +19,7 @@ namespace fmt
 template <>
 struct formatter<VW::string_view> : formatter<fmt::string_view>
 {
-  #if FMT_VERSION >= 90000
+#if FMT_VERSION >= 90000
   template <typename FormatContext>
   auto format(const VW::string_view& sv, FormatContext& ctx) const -> decltype(ctx.out())
 #else

--- a/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
+++ b/vowpalwabbit/core/include/vw/core/vw_string_view_fmt.h
@@ -19,8 +19,13 @@ namespace fmt
 template <>
 struct formatter<VW::string_view> : formatter<fmt::string_view>
 {
+#if FMT_VERSION >= 90000
   template <typename FormatContext>
   auto format(const VW::string_view& sv, FormatContext& ctx) const -> decltype(ctx.out())
+#else
+  template <typename FormatContext>
+  auto format(const VW::string_view& sv, FormatContext& ctx) -> decltype(ctx.out())
+#endif
   {
     return formatter<fmt::string_view>::format({sv.data(), sv.size()}, ctx);
   }

--- a/vowpalwabbit/core/src/action_score.cc
+++ b/vowpalwabbit/core/src/action_score.cc
@@ -7,10 +7,10 @@
 #include "vw/common/string_view.h"
 #include "vw/core/global_data.h"
 #include "vw/core/io_buf.h"
+#include "vw/core/model_utils.h"
 #include "vw/core/text_utils.h"
 #include "vw/core/v_array.h"
 #include "vw/io/logger.h"
-#include "vw/core/model_utils.h"
 
 namespace ACTION_SCORE
 {
@@ -60,8 +60,7 @@ size_t read_model_field(io_buf& io, ACTION_SCORE::action_score& a_s)
   return bytes;
 }
 
-size_t write_model_field(
-    io_buf& io, const ACTION_SCORE::action_score a_s, const std::string& upstream_name, bool text)
+size_t write_model_field(io_buf& io, const ACTION_SCORE::action_score a_s, const std::string& upstream_name, bool text)
 {
   size_t bytes = 0;
   bytes += write_model_field(io, a_s.action, upstream_name + "_action", text);

--- a/vowpalwabbit/core/src/action_score.cc
+++ b/vowpalwabbit/core/src/action_score.cc
@@ -10,6 +10,7 @@
 #include "vw/core/text_utils.h"
 #include "vw/core/v_array.h"
 #include "vw/io/logger.h"
+#include "vw/core/model_utils.h"
 
 namespace ACTION_SCORE
 {
@@ -49,4 +50,23 @@ std::string to_string(const ACTION_SCORE::action_scores& action_scores_or_probs,
   return ss.str();
 }
 
+namespace model_utils
+{
+size_t read_model_field(io_buf& io, ACTION_SCORE::action_score& a_s)
+{
+  size_t bytes = 0;
+  bytes += read_model_field(io, a_s.action);
+  bytes += read_model_field(io, a_s.score);
+  return bytes;
+}
+
+size_t write_model_field(
+    io_buf& io, const ACTION_SCORE::action_score a_s, const std::string& upstream_name, bool text)
+{
+  size_t bytes = 0;
+  bytes += write_model_field(io, a_s.action, upstream_name + "_action", text);
+  bytes += write_model_field(io, a_s.score, upstream_name + "_score", text);
+  return bytes;
+}
+}  // namespace model_utils
 }  // namespace VW

